tools/config/schema_app.json

```diff
@@ -0,0 +1,110 @@
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Mbed Library Schema",
+  "description": "Configuration file for an mbed library",
+  "type": "object",
+  "$id": "http://example.com/root.json",
+  "definitions": {
+    "config_parameter_long": {
+      "type": "object",
+      "properties": {
+        "help": {
+          "description": "An optional help message that describes the purpose of the parameter",
+          "type": "string"
+        },
+        "value": {
+          "description": "An optional field that defines the value of the parameter",
+          "type": [
+            "integer",
+            "string",
+            "boolean"
+          ]
+        },
+        "required": {
+          "description": "An optional field that specifies whether the parameter must be given a value before compiling the code. (False by default)",
+          "type": "boolean"
+        },
+        "macro_name": {
+          "description": "An optional field for the macro defined at compile time for this configuration parameter. The system will automatically figure out the macro name from the configuration parameter, but this field will override it",
+          "type": "string"
+        }
+      }
+    },
+    "config_parameter_short": {
+      "type": [
+        "string",
+        "integer",
+        "boolean"
+      ]
+    },
+    "config_parameter_base": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/config_parameter_long"
+        },
+        {
+          "$ref": "#/definitions/config_parameter_short"
+        }
+      ]
+    },
+    "target_override_entry": {
+      "type": "object",
+      "patternProperties": {
+        "^\\S+$": {}
+      },
+      "additionalProperties": false
+    }
+  },
+  "properties": {
+    "name": {
+      "description": "Name of the library",
+      "type": "string",
+      "items": {
+        "type": "string"
+      }
+    },
+    "config": {
+      "description": "List of configuration parameters",
+      "type": "object",
+      "patternProperties": {
+        "^[^ ]+$": {
+          "$ref": "#/definitions/config_parameter_base"
+        }
+      },
+      "additionalProperties": false
+    },
+    "target_overrides": {
+      "description": "List of overrides for specific targets",
+      "type": "object",
+      "patternProperties": {
+        "\\*": {
+          "type": "object",
+          "patternProperties": {
+            ".*\\..*": {}
+          },
+          "additionalProperties": false
+        },
+        "^\\S+$": {
+          "$ref": "#/definitions/target_override_entry"
+        }
+      },
+      "additionalProperties": false
+    },
+    "macros": {
+      "description": "A list of extra macros that will be defined when compiling a project that includes this library.",
+      "type": "array",
+      "items": {
+        "type": "string",
+        "pattern": "^([A-Za-z0-9_]+|[A-Za-z0-9_]+=[0-9]+|[A-Za-z0-9_]+=\\\".*\\\")$"
+      }
+    },
+    "__config_path": {
```

Comment:
Is this key something that we should allow in the configuration file? or is it something that the tools could add after validation?

Reply:
I wasn't sure when I added it, it was in the allowable keys list, but isn't mentioned anywhere in the specification so it's probably fine to remove.

Reply:
It's added [here](https://github.com/ARMmbed/mbed-os/blob/master/tools/config/__init__.py#L468)

Reply:
You could do the validation before adding that key.

Reply:
That seems sensible, I'll do that.